### PR TITLE
Add OpenMutRef and CloseMutRef

### PR DESCRIPTION
**Pull request by pascal-huber**
- [x] Add OpenMutRef and CloseMutRef when source of assignment is a Deref
 - [x] Add OpenMutRef and CloseMutRef when target of assignment is a Deref

### CODE REVIEW

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 1]
```diff
@@ -840,11 +840,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             }
             mir::Operand::Copy(source) => {
                 let encoded_source = self.encoder.encode_place_high(self.mir, *source)?;
+
+                let mut lifetime_var: Option<vir_high::VariableDecl> = None;
+                if let vir_high::Expression::Deref(vir_high::Deref {
+                    base:
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: `base.get_type()` should give the reference type with the correct `LifetimeConst`.

Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: done, thats much nicer

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 2]
```diff
@@ -821,6 +821,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
+        location: mir::Location,
+        base: &vir_high::Expression,
+        object: vir_high::Expression,
+    ) -> SpannedEncodingResult<()> {
+        if let vir_high::ty::Type::Reference(vir_high::ty::Reference { lifetime, .. }) =
+            base.get_type()
+        {
+            block_builder.add_statement(self.set_statement_error(
+                location,
+                ErrorCtxt::CloseMutRef,
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]:
```suggestion
                ErrorCtxt::OpenMutRef,
```

Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: done

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 3]
```diff
@@ -821,6 +821,44 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
     }
 
+    fn encode_close_mut_ref(
+        &mut self,
+        block_builder: &mut BasicBlockBuilder,
+        location: mir::Location,
+        base: &vir_high::Expression,
+        object: vir_high::Expression,
+    ) -> SpannedEncodingResult<()> {
+        if let vir_high::ty::Type::Reference(vir_high::ty::Reference { lifetime, .. }) =
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: If I understand correctly, this is called only when `base` is known to be a reference, right? If yes, I would suggest adding `else { unreachable!() }` to this if statement.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 4]
```diff
@@ -829,22 +867,54 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         operand: &mir::Operand<'tcx>,
     ) -> SpannedEncodingResult<()> {
         let span = self.encoder.get_span_of_location(self.mir, location);
+
+        let deref_base: Option<vir_high::Expression> = match &encoded_target {
+            vir_high::Expression::Deref(vir_high::Deref { box base, .. }) => Some(base.clone()),
+            _ => None,
+        };
+        if let Some(base) = &deref_base {
+            self.encode_open_mut_ref(block_builder, location, base, encoded_target.clone())?;
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: It seems that in all cases you wrap the calls in `if` statements. I think the code would be more readable if the `if` was inside the function.

**Review by vakaras — APPROVED** [review]
Modulo the comments, looks good to me.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs (hunk 3 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: that is correct, done

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs (hunk 4 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: agree, done

**Review by vakaras — APPROVED** [review]
Looks good. Thanks!

bors r+

**Pull request merged**